text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,673 @@
+that supports most use cases, is sound and fairly intuitive to use. A different approach from a
+type theoretical perspective started out with a much purer scheme, but, when exposed to the
+constraints required, evolved to essentially the same scheme as this RFC. We thus feel confident
+that this RFC is the minimal viable scheme for having bounds on generic parameters of const
+functions. The discussion and evolution of the type theoretical scheme can be found
+[here](https://github.com/rust-rfcs/const-eval/pull/8#issuecomment-452396020) and is only 12 posts
+and a linked three page document long. It is left as an exercise to the reader to read the
+discussion themselves.
+A summary of the result of the discussion can be found at the bottom of [this blog post](https://varkor.github.io/blog/2019/01/11/const-types-traits-and-implementations-in-Rust.html)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It is not a fully general design that supports every possible use case,
+  but it covers the most common cases. See also the alternatives.
+* It becomes a breaking change to add a new method to a trait, even if that method has a default
+  impl. One needs to provide a `const` default impl to not make the change a breaking change.
+* It becomes a breaking change to add a field (even a private one) that has a `Drop` impl which is
+  not `const Drop` (or which has such a field).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## `ConstDrop` trait to opt into const-droppability
+
+Right now it is a breaking change to add a field (even a private one) that has a non-const `Drop`
+impl. This makes `const Drop` a marker trait similar to `Send` and `Sync`. Alternatively we can
+introduce an explicit `ConstDrop` (name bikesheddable) trait, that needs to be implemented for all
+types, even `Copy` types. Users would need to add `T: ConstDrop` bounds instead of `T: Drop` bounds.
+
+## Effect system
+
+A fully powered effect system can allow us to do fine grained constness propagation
+(or no propagation where undesirable). This is out of scope in the near future
+and this RFC is forward compatible to have its background impl be an effect system.
+
+## Fine grained `const` annotations
+
+One could annotate methods instead of impls, allowing just marking some method impls
+as const fn. This would require some sort of "const bounds" in generic functions that
+can be applied to specific methods. E.g. `where <T as Add>::add: const` or something of
+the sort. This design is more complex than the current one and we'd probably want the
+current one as sugar anyway.
+
+## Require `const` bounds everywhere
+
+One could require `const` on the bounds (e.g. `T: const Trait`) instead of assuming constness for all
+bounds. That design would not be forward compatible to allowing `const` trait bounds
+on non-const functions, e.g. in:
+
+```rust
+fn foo<T: const Bar>() -> i32 {
+    const FOO: i32 = T::bar();
+    FOO
+}
+```
```

Comment:
Clarification question: with this RFC, `T: const Trait` is never allowed as a bound, right? The constness of the bound is inferred from the constness of the function, and one can opt-out of that, but that's it.

Reply:
Yes, the `const Trait` bound syntax is kept free with this RFC, so future changes can potentially use it. This RFC solely offers the `?const Trait` syntax

Reply:
Ah, I just noticed this is another alternative listed below ("explicit `const` bounds"). Could you add a link or at least state that explicitly?